frontend/src/utils/lazyRetry.ts

```diff
@@ -0,0 +1,36 @@
+import { ComponentType } from 'react'
+
+/**
+ * Retry React.lazy calls on error, or refresh the page if it was a chunk load
+ * error to hopefully retrieve the correct version of the chunk to due
+ * no-cache header.
+ *
+ * Retrieved from https://www.codemzy.com/blog/fix-chunkloaderror-react.
+ *
+ * @note this only works for route based code splitting https://reactjs.org/docs/code-splitting.html#route-based-code-splitting
+ * Will cause infinite loop if used for normal code splitting since multiple lazy calls may be triggered in a single component.
+ */
+export const lazyRetry = <T extends ComponentType>(
+  componentImport: () => Promise<{ default: T }>,
+) => {
+  return new Promise<{ default: T }>((resolve, reject) => {
+    // check if the window has already been refreshed
+    const hasRefreshed = JSON.parse(
+      window.sessionStorage.getItem('retry-lazy-refreshed') || 'false',
+    )
+    // try to import the component
+    componentImport()
+      .then((component) => {
+        window.sessionStorage.setItem('retry-lazy-refreshed', 'false') // success so reset the refresh
+        resolve(component)
+      })
+      .catch((error) => {
+        if (!hasRefreshed) {
+          // not been refreshed yet
+          window.sessionStorage.setItem('retry-lazy-refreshed', 'true') // we are now going to refresh
+          return window.location.reload() // refresh the page
```

Comment:
If there is a client-side cache issue, this soft refresh might not be enough to get the latest version 🤔

We could try a more aggressive refresh with a dummy query string arg, like `?r=${Math.random()`

Reply:
it should, since our `index.html` has no caching, so a refresh should refetch a fresh `index.html`... in theory.

Reply:
The way I was looking at this is like that: if the links are broken now, we're assuming a bad cache. Something was cached incorrectly for some reason (??). If that incorrect cache is expiry based, the browser might still use whatever copy it holds on a refresh, without doing a full reload.

This is a bit of stretch, with a no-cache policy header, there should not be random expiry based cache happening, so I'm good that a reload should fetch the right thinngs.

That being said... are there ways other than cache we can get 404s on chunks? And I'm thinking maybe yes 😢 . Splitting the app in chunks means the client makes more requests to fetch assets, and at release time we still have both old and new version servers running in parallel, so some requests for new chunks could hit old servers (or vice versa) and fail.

During deployments, there could even be consecutive failures depending on which servers the requests hit. Urgh 😢 

And if we eventually do extended blue/green deployments, that problem would only get worse. We'll have to address it at some point.